plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableMetaQuery.php

```diff
@@ -106,6 +106,10 @@ public function __construct( OrdersTableQuery $q ) {
 			return;
 		}
 
+		if ( $this->is_atomic( $meta_query ) ) {
+			$meta_query = array( $meta_query );
```

Comment:
Question ✋🏽 ...

Suppose we have a case now or in the future where the query args are filtered **before** being passed to `OrdersTableMetaQuery`, and let's suppose the meta query initially looks like this (ie, it uses the 'simple format' that we are adding support for in this PR):

```php
[
	'key'   => 'foo',
	'value' => 'bar',
]
```

Then a filter function appends a further meta query in the traditional style, so we end up with something like:

```php
[
	'key'   => 'default',
	'value' => '123',
	[
		'key'   => 'filter-tweak',
		'value' => '456',
	],
]
```

When execution reaches this point, the `is_atomic()` check will be satisfied (because `isset( $arg['key'] )`) and we will then wrap the entire structure, giving us:

```php
[
	[
		'key'   => 'default',
		'value' => '123',
		[
			'key'   => 'filter-tweak',
			'value' => '456',
		],
	],
]
```

I didn't actually test this, but I suspect this means only the default component of the meta query would be respected and the additional meta query added by the filter function will not be 'seen', because it's nested one level too far, which seems 'unfair' and could lead to unexpected behavior.

(We could imagine the reverse happening, too: where the default and all other filter functions follow the WP norm of nested arrays—then a single filter function decides to use the simpler format, which would effectively mask all those other meta queries.)

I think we can mitigate this risk (assuming what I described above sounds plausible 😅) by moving the top-level meta query args, if they exist into a sub-array of their own. But, I still wonder if it is ideal: could it still be confusing for filter functions where they are examining existing meta queries before deciding to make a change (since the correction to the standard nested array structure would not happen until execution reaches this class)?

Hope that all makes sense!

Reply:
@barryhughes: Thanks for this wonderful analysis! I think you're right, and I might've gotten a bit ahead of myself here thinking `WP_Query` did allow the `meta_query` to be in this "simpler" format. I did some testing, and actually `WP_Query` totally ignores `key` and `value` if defined outside of an array (i.e. if `meta_query` is not an array of arrays). For example, the following code will only look for posts matching `other_key = other_value` and will ignore `my_key` entirely:

```php
$q = new WP_Query(
	[
		'meta_query' =>
		[
			'key'   => 'my_key',
			'value' => 'my_value',
			[
				'key'   => 'other_key',
				'value' => 'other_value',
			]
		]
	]
);
```

I believe this in fact "correct behavior". This doesn't necessarily prevent devs from abusing filters to produce incorrect `meta_query` arrays, but at least means the expectation is that you shouldn't be using the simple/direct format at all.

What is different is that WP just silently ignores things, and we're throwing a fatal error. Maybe we should just trigger a `doint_it_wrong` or also silently ignore. What do you think?

Reply:
> I believe this in fact "correct behavior".

Yep, I agree.

> What is different is that WP just silently ignores things, and we're throwing a fatal error. Maybe we should just trigger a doint_it_wrong or also silently ignore.

Since our COT and CPT implementations share in common at least one "port of entry" in the form of `wc_get_orders()`, and since the caller arguably doesn't know/shouldn't have to care which data-store is in use, I'd personally err on the side of matching WordPress and silently ignoring (or perhaps at most _logging)._

I also believe there are plugins out there that pass non-standard query arg keys to `WP_Query` (and then use hooks to detect their presence and modify the query as needed). Though I have no idea if this is common in relation to the `meta_query` specifically, my more general point is just that being overly strict here could lead to friction.

With that said, if we wanted to be stricter/noisier on encountering unexpected values, this might be a good stage to do that and test the waters (since COT is not quite at production-readiness yet). Personally, though, I'd probably run with the silently ignore strategy 🙃

Reply:
Thanks Barry! I've updated the code based on this discussion.